NIFI-4215 Fixed stackoverflow error when NiFi tries to parse an Avro record that has a field of the same record type

Wesley-Lawrence opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [✓] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [✓] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [✓] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [✓] Is your initial contribution a single, squashed commit?

### For code changes:
- [✓] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [✓] Have you written or updated unit tests to verify your changes?
- [N/A] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [N/A] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [N/A] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [N/A] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [N/A] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

jvwing commented:
Reviewing...

jvwing commented:
@Wesley-Lawrence Thanks for putting this PR together.  I am still reviewing - the only immediate feedback I can give is that I would prefer not to update the checkstyle definition in pom.xml as part of this fix for the AvroSchemaRegistry issue.

Wesley-Lawrence commented:
Thanks for taking a look @jvwing!

I didn't want to change it, but I keep getting the following error with it;
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:2.15:check (check-style) on project nifi-avro-record-utils: You have 1 Checkstyle violation. -> [Help 1]
```

Earlier in the maven log;
```
[INFO] --- maven-checkstyle-plugin:2.15:check (check-style) @ nifi-avro-record-utils ---
[WARNING] src/main/java/org/apache/nifi/avro/AvroTypeUtil.java[275:17] (blocks) RightCurly: '}' should be on the same line.
```

Which references a `}` I added here;
```
273                if (knownRecordTypes.containsKey(schemaFullName)) {
274                    return knownRecordTypes.get(schemaFullName);
275            --> }
276                else {
```

However, this is the style used everywhere in NiFi, and is the one defined by the `RightCurly` section below the one I removed.

I suspect it's something weird in my environment, but removing the default `RightCurly` definition fixed my issue, and it looks like it was just left over from some old migration, so I figured it could be safely removed.

Out of curiosity, if you run a contrib check, do you get the same error I do?

pvillard31 commented:
@Wesley-Lawrence I believe this should be:

````java
if() {

} else {

}
````

Wesley-Lawrence commented:
@pvillard31 Yup, you're right, that solves the issue for me 😃.

I got so caught up in the second `RightCurly` definitons saying `}` should be alone, and other `if`s being that way, I didn't see the correct style for `if ... else`s.

I'll fix that line, re-add the original `RightCurly` definition back in, and push a new squashed commit.

Thanks again, both of you.

jvwing commented:
Thanks @pvillard31!

@Wesley-Lawrence Don't worry about squashing yet, we can do that as a final step.

jvwing commented on nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java:
```diff
@@ -218,6 +218,15 @@ private static Schema nullable(final Schema schema) {
+    public static DataType determineDataType(final Schema avroSchema, Map<String, DataType> knownRecordTypes) {
+
+        if (knownRecordTypes == null) {
+            throw new IllegalArgumentException("'foundSchemas' cannot be null.");
```
Did you mean this to say "knownRecordTypes" rather than "foundSchemas"?

jvwing commented:
@Wesley-Lawrence  I think this is a good fix, your approach to the solution seems solid.  Thanks for adding the unit tests for the recursive and mutually-referential cases.  Would you please:

1. Fix the checkstyle issue and remove the change to the checkstyle definitions
1. Optionally change the foundSchemas/knownRecordTypes exception message
1. Squash and rebase on master

Wesley-Lawrence commented:
Alright, changes made, and contrib-check still passing for me.

Thanks for the review @jvwing!

jvwing commented:
Looks good.  Thanks again for fixing this, @Wesley-Lawrence.
